src/components/containers/Panel.js

```diff
@@ -1,51 +1,141 @@
-import React, {Component} from 'react';
+import PanelHeader from './PanelHeader';
 import PropTypes from 'prop-types';
+import React, {Component} from 'react';
+import update from 'immutability-helper';
 import {bem} from 'lib';
-import {ChartLineIcon} from 'plotly-icons';
-class Panel extends Component {
-  render() {
-    const {visible, children} = this.props;
 
-    if (visible) {
-      return <div className={bem('panel')}>{children}</div>;
-    }
-    return null;
+export default class Panel extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      nbOfFolds: 0,
+      individualFoldStates: [],
+    };
+    this.toggleFolds = this.toggleFolds.bind(this);
+    this.toggleFold = this.toggleFold.bind(this);
   }
-}
 
-export class PanelHeader extends Component {
-  render() {
-    const {children, action} = this.props;
+  getChildContext() {
+    return {
+      individualFoldStates: this.state.individualFoldStates,
+      toggleFold: this.toggleFold,
+    };
+  }
 
+  toggleFolds() {
+    const {individualFoldStates} = this.state;
+    const hasOpen =
+      individualFoldStates.length > 0 &&
+      individualFoldStates.some(s => s !== true);
+    this.setState({
+      individualFoldStates: individualFoldStates.map(() => {
+        if (hasOpen) {
+          return true;
+        }
+        return false;
+      }),
+    });
+  }
+
+  toggleFold(index) {
+    this.setState(
+      update(this.state, {
+        individualFoldStates: {
+          [index]: {$set: !this.state.individualFoldStates[index]},
```

Comment:
Here I would use `$toggle` instead of `$set` like this: `update(this.state, {individualFoldStates: {$toggle: [index]}})`

Reply:
oh, thanks :)

Reply:
I know the docs say that `$toggle` works on objects but this works, I just tried it